lsp-fiber/src/rpc.ml

```diff
@@ -190,19 +184,31 @@ struct
         Fiber.return
           (Jsonrpc_fiber.Reply.now (Jsonrpc.Response.error req.id error), state)
       | Ok (In_request.E r) ->
+        let cancel = Fiber.Cancel.create () in
         let+ response, state =
-          let cancel = Fiber.Cancel.create () in
-          Fiber.finalize
-            (fun () ->
-              Fiber.Var.set cancel_token cancel (fun () ->
-                  Table.replace t.pending req.id cancel;
-                  h_on_request.on_request t r))
-            ~finally:(fun () ->
-              Table.remove t.pending req.id;
-              Fiber.return ())
+          Fiber.Var.set cancel_token cancel (fun () ->
+              Table.replace t.pending req.id cancel;
+              h_on_request.on_request t r)
+        in
+        let to_response x =
+          Jsonrpc.Response.ok req.id (In_request.yojson_of_result r x)
         in
         let reply =
-          Reply.to_jsonrpc response req.id (In_request.yojson_of_result r)
+          match response with
+          | Reply.Now r ->
+            Table.remove t.pending req.id;
+            Jsonrpc_fiber.Reply.now (to_response r)
+          | Reply.Later k ->
+            let f send =
+              Fiber.finalize
+                (fun () ->
+                  Fiber.Var.set cancel_token cancel (fun () ->
+                      k (fun r -> send (to_response r))))
+                ~finally:(fun () ->
+                  Table.remove t.pending req.id;
```

Comment:
Removing that line here and also in the `Reply.now` case still makes the tests pass. Should we try to have a test to cover that we correctly remove the token from the `pending` table? If so, can we do it easily with the current public API?

Reply:
I don't how to do it. Even if we don't cleanup the table, all we're doing is wasting memory so it isn't the biggest deal. I'd say not to worry about it for now.